src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java

```diff
@@ -865,17 +1093,106 @@ public static FormValidation doCheckBitbucketServerUrl(@QueryParameter String bi
             return FormValidation.ok();
         }
 
-        public ListBoxModel doFillCredentialsIdItems(@AncestorInPath SCMSourceOwner context, @QueryParameter String bitbucketServerUrl) {
+        public static FormValidation doCheckServerUrl(@QueryParameter String value) {
+            if (BitbucketEndpointConfiguration.get().findEndpoint(value) == null) {
+                return FormValidation.error("Unregistered Server: " + value);
+            }
+            return FormValidation.ok();
+        }
+
+        public boolean isServerUrlSelectable() {
+            return BitbucketEndpointConfiguration.get().isEndpointSelectable();
+        }
+
+        public ListBoxModel doFillServerUrlItems() {
+            return BitbucketEndpointConfiguration.get().getEndpointItems();
+        }
+
+        public ListBoxModel doFillCredentialsIdItems(@AncestorInPath SCMSourceOwner context, @QueryParameter String serverUrl) {
             StandardListBoxModel result = new StandardListBoxModel();
             result.includeEmptyValue();
-            return BitbucketCredentials.fillCredentials(bitbucketServerUrl, context, result);
+            result.includeMatchingAs(
```

Comment:
There are only two places currently where this exact code should live: the navigator and the source.

Now at some point in time it is conceivable to introduce a domain specification that includes the bitbucket repository, in which case the *currently* duplicate code would cease to be so.

I have learned through experience that duplicate code is bad but deduplication of business logic is not always so. In fact consolidating closely related but different business logic just because it looks the same can then result in unintended consequences when changes are made.

My considered risk assessment for this code was de-duplication is incorrect and hence why I inlined the method